#!/bin/bash
#SBATCH --partition main
#SBATCH --nodes 1
#SBATCH --ntasks 10
#SBATCH --time 01:00:00
#SBATCH --mem 4g
#SBATCH --account=ttrojan_001
#SBATCH --chdir /home1/ttrojan/computational-biology-on-carc
module purge
module load usc
module load sratoolkit
module load bwa
module load samtools
module load bcftools
module load vcftools

# setuo sra toolkit
vdb-config --restore-defaults

# create outpout directory
outdir="results/variant-calling"
mkdir -p ${outdir}

sra_list="SRR2584866 SRR2589044"
# set up the reference genome and what files need to be generated by the index job
reference_genome=data/ecoli_rel606.fasta
# index the reference file
bwa index ${reference_genome}

for sra in ${sra_list}
do
  # download SRA
  fasterq-dump --split-files -e 1 ${sra}
  # align reads tp reference genome job
  bwa mem -t 10 ${reference_genome} ${sra}_1.fastq ${sra}_1.fastq > ${outdir}/${sra}.sam
  # Aligning and converting $sam to bam format $bam
  samtools view -S -b ${sra}.aligned.sam > ${outdir}/${sra}.aligned.bam
  # Sorting $bam file to $sorted_bam
  samtools sort -o ${outdir}/${sra}.aligned.sorted.bam ${outdir}/${sra}.aligned.bam
  # Indexing $sorted_bam file
  samtools index ${outdir}/${sra}.aligned.sorted.bam
  # Variant calling
  # bcftools for calculating the read coverage of positions in the genome
  bcftools mpileup --threads 10 -O b -o ${outdir}/${sra}.raw.bcf -f ${reference_genome} ${outdir}/${sra}.aligned.sorted.bam
  # bcftools for Detect the single nucleotide variants (SNVs)
  bcftools call --ploidy 1 -m -v -o ${outdir}/${sra}.vcf ${outdir}/${sra}.raw.bcf
  # vcfutils Filter and report the SNV variants in variant calling format (VCF)
  vcfutils.pl varFilter ${outdir}/${sra}.vcf > ${outdir}/${sra}_final.vcf
done


